### PR TITLE
Upgrade to React 18.3

### PR DESCRIPTION
**Pull request by gjvoosten**
Migrate our React code to TypeScript (replacing propTypes and defaultProps), and upgrade React to 18.3.

Closes [AB#1102](https://dev.azure.com/ncia-anet/2aa083a5-af3d-44e1-8c7b-6e9e6b124d91/_workitems/edit/1102)

#### User changes
- None.

#### Superuser changes
- None.

#### Admin changes
- None.

#### System admin changes
- [ ] application.yml or anet-dictionary.yml needs change
- [ ] db needs migration
- [ ] documentation has changed
- [ ] graphql schema has changed

### Checklist
- [x] described the user behavior in PR body
- [x] referenced/updated all related issues
- [x] commits follow a `repo#issue: Title` title format and [these 7 rules](https://chris.beams.io/posts/git-commit/)
- [x] commits have a [clean history](https://epage.github.io/dev/commits/), otherwise PR may be squash-merged
- [ ] added and/or updated unit tests
- [ ] added and/or updated e2e tests
- [ ] added and/or updated data migrations
- [ ] updated documentation
- [x] resolved all build errors and warnings
- [x] opened debt issues for anything not resolved here

### CODE REVIEW

File: client/src/HOC/LoaderHOC.tsx [hunk 1]
```diff
@@ -8,8 +7,12 @@ export const mapDispatchToProps = (dispatch, ownProps) => ({
   hideLoading: () => dispatch(hideLoading())
 })
 
+interface WrapperProps {
+  loaderMessage?: string | any
```
Comment (midmarch) [inline client/src/HOC/LoaderHOC.tsx:11]: Redundant, same as `any`

File: client/src/components/AdvisorReports/AdvisorReportsRow.tsx [hunk 2]
```diff
@@ -22,7 +21,16 @@ const _advisorStats = (columnGroups, statistics) => {
...
 }
 
-const AdvisorReportsRow = props => {
+interface AdvisorReportsRowProps {
+  checked?: boolean
+  columnGroups?: number[]
+  withOrganizationLink?: boolean
+  link?: any
+  onSelectRow?: (...args: unknown[]) => unknown
+  row?: any
```
Comment (midmarch) [inline client/src/components/AdvisorReports/AdvisorReportsRow.tsx:30]:
```suggestion
  link?: React.ReactNode
  onSelectRow?: (event: React.ChangeEvent<HTMLInputElement>) => void
  row?: { stats?: any; name?: string }
```

**Review by midmarch — APPROVED** [review]
Porbably it does not make sense yet to see if typings can be improved, as we need start somewhere.

I have no problems merging it.

**Comment by gjvoosten** [comment]
> Porbably it does not make sense yet to see if typings can be improved, as we need start somewhere.
> 
> I have no problems merging it.

Indeed; I kept the typings mostly (or at least as close as possible to) how they were generated by the codemod. The typings can definitely be improved (as shown by e.g. the lint warnings). I created a debt issue in AB#1203 for that.

**Pull request merged**